### DISCUSSION TITLE
Is it possible not to create UI on start but serve only as web server?

### DISCUSSION THREAD

**Discussion by pashagolub** [wailsapp/wails#1635] [Q&A] [answered]
Hello,

I tried hard to find the answer myself but failed. I want to add UI to one of my Go applications. The thing is, it's widely run in Docker containers, and in this case, there is no reason to create any windows. The only solution is to serve UI in a browser. 

Can `wails` build an application that serves as a web server only, without creating window/webview/UI?

**Comment by leaanthony** [comment, marked as answer]
That's something we are planning and most of it is already there to be honest. @tmclane is keen on this too.

**Reply by tmclane** [reply]
I have the code ready and almost mergeable.  I am just not happy with it at the moment hence the delay. I think this gives me the extra push to finish it and open the PR.

**Reply by pashagolub** [reply]
Thanks a lot! That would be awesome!

**Reply by pashagolub** [reply]
Congrats with v2! Any news on this? No pressure! Just asking :)

**Reply by alienator88** [reply]
Hi! Any news on this? I also have a good use case for displaying the app within a browser as well.

**Reply by leaanthony** [reply]
There's been a lot of work recently on adding request handlers so we need to review the draft PR. What are your use cases?

**Reply by alienator88** [reply]
I have an internal tool which needs to execute shell and other CLI tool commands locally on each user's computer, which works fine as a desktop app. Would like to convert this to a "website" in a sense if possible with the Wails app running and serving the app in the browser for the user. Because of other internal constraints, this would be the only way for me to have the app run in a browser.

If possible in the future, maybe even not launching a window at all and just having a menubar item, whenever that would be available. I think I saw something like that on the roadmap.

**Reply by leaanthony** [reply]
The last one is certainly something we are planning on doing

**Reply by pashagolub** [reply]
Sorry, missed the question. My [application](https://github.com/cybertec-postgresql/[pg_timetable](https://github.com/cybertec-postgresql/pg_timetable)) is a database scheduler which usually runs beside the PostgreSQL server or in a docker container somewhere nearby. But not as a client application. I want to provide users with a dashboard where they can create/edit/delete jobs, start/pause/stop jobs, see some basic stats.

**Reply by oSethoum** [reply]
@pashagolub I don't see why you would choose wails for such a use case; you can serve the UI (HTML/CSS/JS) from web server using either `net/http` package or other web frameworks such as `fiber`.

**Reply by pashagolub** [reply]
oSethoum, thanks for reply. I want to have both of worlds :-) Also I had experience in developing GUI apps and zero in web apps. So I thought it would be easier to use wails co create full GUI and only then to use it as a web app without modifying the code base

**Reply by oSethoum** [reply]
@pashagolub you are welcome, my knowledge with wails is limited but in this use case I don't think wails would be the best option, even if you are planning to use wails somehow you still need to make a web app, checkout how does wails work [here](https://wails.io/docs/howdoesitwork/) `frontend assets` is basically a web app.

**Reply by Shackelford-Arden** [reply]
Just to make sure I'm following the correct things, is this the PR we're following for this discussion, @tmclane ?

https://github.com/wailsapp/wails/pull/1652

**Comment by FrenchGithubUser** [comment]
I'm also pretty interested in this, is it somewhere in the roadmap ?

**Comment by pkuntik** [comment]
me too

**Comment by marcio199226** [comment]
i'm also interested, this will be a game changer for features that are not supported by webview itself of by the specific platform webview implementation such us lack of webrtc/mediaDevices on linux reported by @oxzjh [here](https://github.com/wailsapp/wails/discussions/2697)

I have some trouble now because I can't implement voicecall on linux, but if I could run wails only as backend (like waisl dev works) and call golang code from websocket that would be awesome

**Reply by leaanthony** [reply]
I'm curious as to why Wails would be considered for non-desktop scenarios over just plain Go using http. Genuine question that will help drive this feature 👍

**Reply by FrenchGithubUser** [reply]
this allows to have both (web service and standalone app) with only one codebase (at least in my case). For people who want to use the program on a single computer or as a web service on their local network for example

**Reply by marcio199226** [reply]
yeah I also thought about this use case: simply we are not tied to platform specific webview implementation limitation

`I'm curious as to why Wails would be considered for non-desktop scenarios over just plain Go using http`
The first use case as stated by @pashagolub in the first post then as I wrote above we are not tied to platform's webview but user can use any chromium based browser on his system, so we as web developers have the latest web browser api availables :)

I will try to explain my point of view, we can develop a desktop app (the be part) which is cross platfrom at the same time using the latest web technologies because our app is not tied to any webview (if developer choose that type of build type)

The advantages of not **"using http"** for me it's that developing in the current flow seems really native from perspective of an web developer and wails provide all stuff for ipc that feels more natural way and fast over a http implementation with standard req/res architecture

I don't know if my arguments are enough

is this somehow similar to what this  #1652 implements ?

**Reply by FrenchGithubUser** [reply]
Yes I think it is what this PR implements and there has been recent activity on it !

**Comment by kmahyyg** [comment]
Most application with Web-based UI in such kind of way, can allow user to deploy application on remote infra and access UI via browser (which is already installed on almost every workstation today), and dev won't have to write another UI or open RDP, bring both slim binary and easy to access with some security.

As of 2026, may I ask if there's any way to achieve this? Thanks a lot for your excellent work.

**Reply by symball** [reply]
Very happy to be wrong about this and we now have server mode, awesome

~~Here is word on this matter straight from the creator~~

~~We always recommend developing using the native app instead of the browser for a simple reason: you may end up using APIs that are unsupported in webviews. We've had so many issues opened where "it works in Dev but not in prod builds" that we've dropped support for browser based Dev in v3. You can still connect to vite to do UI updates but the bindings won't work.~~

~~In short, Wails chooses to focus on working well within webkit / the way bindings operate, and is not really a priority.~~

~~There is a relatively simple way to achieve what you want depending on your architecture if you don't need to use the wails runtime specific features. The bindings are not really much different from a typical API so, if you have bindings like the following example~~

```
func (i *ToolInterface) ToolStateGet() *tooldto.ToolState {
    return i.toolService.State()
}

func (i *ToolInterface) ToolGetAll() []*tooldto.Tool {
    return i.toolService.GetAllTools()
}

func (i *ToolInterface) ToolGetRecord(toolRef tooldto.ToolRef) (*tooldto.Tool, error) {
    return i.toolService.GetToolRecord(toolRef)
}
```

~~it would be minimal effort to turn the interface in to an actual API served via http. at that point, you would fire up the backend / frontend dev server yourself. To make the process seamless, have two main entry points to your app that get chosen depending on build tags.~~

```
// main_desktop.go
//go:build desktop

...fires up wails...

// main_browser.go
//go:build browser

...fires up API
```

~~On the frontend, have an abstraction layer to differentiate between using the wails bindings or your api.~~

~~These are some principles of clean code i would highly recommend regardless; relying on interfaces instead of solid implementations makes it easy to mock certain features in your test environment~~

**Comment by leaanthony** [comment]
Good news: **this is fully supported in Wails v3** via the experimental server plugin.

Wails v3 lets you compile the same codebase as either a desktop app **or** a headless HTTP server using a build tag. No native window is created in server mode.

**Setup**

Add the server plugin to your app options:

```go
import (
    "github.com/wailsapp/wails/v3/pkg/application"
    "github.com/wailsapp/wails/v3/plugins/experimental/server"
)

func main() {
    app := application.New(application.Options{
        Plugins: map[string]application.Plugin{
            "server": server.NewPlugin(&server.Config{
                Host: "0.0.0.0",
                Port: 31115,
            }),
        },
    })

    app.Run()
}
```

**Build for server/headless mode**

```bash
go build -tags server -ldflags="-s -w" -o myapp-server .
```

Without  you get the normal desktop binary. With it, the native WebView is skipped entirely and the app serves its frontend over HTTP — ideal for Docker/server deployments.

- Frontend → Backend: HTTP requests
- Backend → Frontend: WebSocket broadcast

See the [v3 server-build guide](https://v3alpha.wails.io/guides/server-build/) for full details.

For v2 this is a known limitation — the window is mandatory. If you're on v2 and this is blocking you, the [v2 → v3 migration guide](https://v3alpha.wails.io/migration/v2-to-v3/) shows it typically takes 1–4 hours.
